### PR TITLE
Implement call

### PR DESCRIPTION
**Pull request by axic**
Depends on #47.

### CODE REVIEW

File: lib/fizzy/execute.cpp [hunk 1]
```diff
@@ -229,6 +229,31 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
             break;
         case Instr::end:
             goto end;
+        case Instr::call:
+        {
+            const auto idx = read<uint32_t>(immediates);
+            assert(idx < instance.module.typesec.size());
+            auto inputs = instance.module.typesec[idx].inputs.size();
```
Comment (gumb0) [inline lib/fizzy/execute.cpp]: I don't think this is correct, immediate of `call` is the function index, not the type index... so we need to parse function section for this

File: lib/fizzy/execute.cpp [hunk 2]
```diff
@@ -229,6 +229,31 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
+            const auto outputs = instance.module.typesec[idx].outputs.size();
+            // FIXME: optimise this
+            uint64_stack cargs;
```
Comment (gumb0) [inline lib/fizzy/execute.cpp]: I would try
```
uint64_stack cargs(stack.rbegin(), stack.rbegin() + inputs.size());
stack.erase(stack.end() - inputs.size(), stack.end());
```

File: lib/fizzy/execute.cpp [hunk 3]
```diff
@@ -229,6 +229,35 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
             break;
         case Instr::end:
             goto end;
+        case Instr::call:
+        {
+            const auto func_idx = read<uint32_t>(immediates);
```
Comment (gumb0) [inline lib/fizzy/execute.cpp]: Note: this will need to be adjusted when imported functions are supported.

File: lib/fizzy/execute.cpp [hunk 4]
```diff
@@ -229,6 +229,35 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
+            stack.erase(stack.end() - static_cast<ptrdiff_t>(inputs), stack.end());
+
+            const auto ret = execute(instance, func_idx, cargs);
+            // Bubble up traps
+            if (ret.trapped)
+            {
+                trap = true;
+                goto end;
+            }
+            const auto outputs = instance.module.typesec[type_idx].outputs.size();
```
Comment (chfast) [inline lib/fizzy/execute.cpp]:
```suggestion
            const auto num_outputs = instance.module.typesec[type_idx].outputs.size();
```

File: lib/fizzy/execute.cpp [hunk 5]
```diff
@@ -229,6 +229,35 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
+            const auto func_idx = read<uint32_t>(immediates);
+            assert(func_idx < instance.module.funcsec.size());
+            const auto type_idx = instance.module.funcsec[func_idx];
+            assert(type_idx < instance.module.typesec.size());
+            const auto inputs = instance.module.typesec[type_idx].inputs.size();
+
+            assert(stack.size() >= inputs);
+            std::vector<uint64_t> cargs(
+                stack.rbegin(), stack.rbegin() + static_cast<ptrdiff_t>(inputs));
+            stack.erase(stack.end() - static_cast<ptrdiff_t>(inputs), stack.end());
```
[the remark this answers is not in the document]
Comment (axic) [reply lib/fizzy/execute.cpp]: @gumb0 any comment?

[the remark this answers is not in the document]
Comment (gumb0) [reply lib/fizzy/execute.cpp]: Yeah, should work, let's change it

[the remark this answers is not in the document]
Comment (axic) [reply lib/fizzy/execute.cpp]: Ok, doing it.

File: lib/fizzy/execute.cpp [hunk 6]
```diff
@@ -229,6 +229,35 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
+        {
+            const auto func_idx = read<uint32_t>(immediates);
+            assert(func_idx < instance.module.funcsec.size());
+            const auto type_idx = instance.module.funcsec[func_idx];
+            assert(type_idx < instance.module.typesec.size());
+            const auto inputs = instance.module.typesec[type_idx].inputs.size();
+
+            assert(stack.size() >= inputs);
+            std::vector<uint64_t> cargs(
+                stack.rbegin(), stack.rbegin() + static_cast<ptrdiff_t>(inputs));
```
[the remark this answers is not in the document]
Comment (axic) [reply lib/fizzy/execute.cpp]: Since now it is a single cast and since it breaks the compilation of the assert, I'm keeping it as is.

**Pull request merged**